### PR TITLE
[Feature](bangc-ops): add new op moe_dispatch_backward_gate

### PR DESCRIPTION
**Pull request by wushaoqiang2168**
## 1. Motivation

新算子moe_diaptch_backward_gate.

## 2. Modification

新算子moe_diaptch_backward_gate.

- bangc-ops/kernel_depends.toml
- bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp
- bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu
- bangc-ops/kernels/utils/common.h
- bangc-ops/mlu_op.h
- bangc-ops/mlu_op_kernel.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/test_case/case_0.prototxt
- docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md
- docs/bangc-docs/user_guide/9_operators/index.rst

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- 本算子属于累加类算子，验收标准为 diff1 <= 3e-3 && diff2 <= 3e-3。

#### 3.1.2 Operator Scheme checklist

| No.  | Details                                | Check Results                                                |
| ---- | -------------------------------------- | ------------------------------------------------------------ |
| 1    | Supported hardware                     | MLU370 <br />MLU590                                          |
| 2    | Job types                              | U1                                                           |
| 3    | Layouts                                | ARRAY                                                        |
| 4    | Whether multi-dimensions are supported | no                                                           |
| 5    | Whether element zero is supported      | yes                                                          |
| 6    | Data type(half/float)                  | float                                                        |
| 7    | Whether there is size limit            | indices_desc、locations_desc、input_desc、grad_gates_desc的第一个维度大小等于samples <br />input_desc、dispatch_desc的第2个维度大小等于hidden <br />dispatch_desc的第1个维度大小等于num_experts * capacity |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x]  Data type test
- [ ]  Multi-dimensional tensor test
- [ ]  Layout test
- [x]  Different size/integer remainder end segment/alignment misalignment test
- [x]  Zero dimensional tensor test/zero element test
- [x]  stability test
- [x]  Multiple platform test
- [x]  Gen_case module test
- [x]  Nan/INF tests
- [ ]  Bug fix tests
- [x]  For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x]  For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x]  For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point                                      | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | ------------------- | --------------------------- |
| Whether it conforms to the operator restriction | Normal error        | 通过                            |
| Whether illegal parameters are passed           | Normal error        |     通过                        |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

| Test Point         | Description                     | Quantity | Comment                                                      |
| ------------------ | ------------------------------- | -------- | ------------------------------------------------------------ |
| Data type test     | float                           | 通过     | float<br />[ SUMMARY  ] Total 580 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 580 test cases from 1 test suite ran. (13244 ms total)<br/>[  PASSED  ] 580 test cases. |
| Zero element test  | Whether to support this test    | 通过     | samples=0<br />[MLUOP] [Vlog]:[MoeDispatchBackwardGateExecutor] call compute() begin.<br/>[MLUOP] [Vlog]:[mluOpMoeDispatchBackwardGate] Skip zero element tensor.<br/>[MLUOP] [Vlog]:[MoeDispatchBackwardGateExecutor] call compute() end.<br /><br />hidden=0<br />[MLUOP] [Vlog]:[mluOpMoeDispatchBackwardGate] Skip zero element tensor.<br/>[MLUOP] [Vlog]:mluopFill start.<br/>[MLUOP] [Vlog]:mluopFill end.<br /><br />capacity=0<br />[MLUOP] [Vlog]:[mluOpMoeDispatchBackwardGate] Skip zero element tensor.<br />[MLUOP] [Vlog]:mluopFill start.<br/>[MLUOP] [Vlog]:mluopFill end.<br /><br />num_experts=0<br />[MLUOP] [Vlog]:[mluOpMoeDispatchBackwardGate] Skip zero element tensor.<br />[MLUOP] [Vlog]:mluopFill start.<br/>[MLUOP] [Vlog]:mluopFill end. |
| Stability test     | --gtest_repeat=NUM --thread=NUM | 通过     | ./mluop_gtest --gtest_filter="moe_dispatch_backward_gate*" --cases_dir=../../../../gtest_case/moe_dispatch_backward_gate/ --gtest_repeat=100 --thread=100<br />[ SUMMARY  ] Total 58000 cases of 100 op(s).<br/>ALL PASSED.<br/>[==========] 1 test case from 1 test suite ran. (2101 ms total)<br/>[  PASSED  ] 1 test case. |
| Mult-platform test | MLU370/MLU590                   | 通过     | MLU370:<br />[ SUMMARY  ] Total 580 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 580 test cases from 1 test suite ran. (13244 ms total)<br/>[  PASSED  ] 580 test cases.<br />MLU590<br />[ SUMMARY  ] Total 580 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 580 test cases from 1 test suite ran. (56126 ms total)<br/>[  PASSED  ] 580 test cases. |
| Nan/INF test       | Whether to support this test    | 通过     | ./mluop_gtest --gtest_filter="moe_dispatch_backward_gate*" --cases_dir=../../../../gtest_case/moe_dispatch_backward_gate/naninf/<br />ALL PASSED.<br/>[==========] 64 test cases from 1 test suite ran. (11056 ms total)<br/>[  PASSED  ] 64 test cases. |
| 防呆测试           | 输入参数防呆测试                | 通过     | 1. 指针为NULL测试：handle、indices_desc、 locations_desc、input_desc、dispatch_desc、grad_gates_desc、indices、locations、input、dispatch、grad_gates、workspace 测试均通过，正确报错，如<br />[MLUOP] [Error]:[mluOpMoeDispatchBackwardGate] Check failed: handle != NULL.<br /><br />2.shape / dim测试<br />[MLUOP] [Error]:[mluOpMoeDispatchBackwardGate] Check failed: (samples == indices_desc->dims[0])<br />[MLUOP] [Error]:[mluOpMoeDispatchBackwardGate] Check failed: grad_gates_desc->dim == 1.<br /><br />[MLUOP] [Error]:[mluOpMoeDispatchBackwardGate] Check failed: ((num_experts * capacity) == dispatch_desc->dims[0]).<br /><br />3.large tensor测试<br />[MLUOP] [Warning]:[mluOpSetTensorDescriptor]: overflow max tensor num.Currently, mluOp supports tensor num smaller than 2^31,now tensor dims:(3500,51201200), data_width:4.<br /><br />4.dtype测试<br />[MLUOP] [Error]:[mluOpMoeDispatchBackwardGate] Check failed: (locations_desc->dtype == MLUOP_DTYPE_INT32). Only int32 are supported in locations tensor, but the data type of tensor is DTYPE_FLOAT. |
| 代码覆盖率         | 测试结果                        | 通过     | Filename [Sort by name]               line_coverage]           by_function<br/>                                                               coverage]<br/>moe_dispatch_backward_gate_union1.mlu [95.4%] 95.4 %310 / 325 50.0 %7 / 14<br/>                                      [95.4%] |
| 内存泄漏测试       | 测试结果                        | 通过     |                                                              |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU590

| Operation                  | Mlu_hardware_time(us) | Mlu_interface_time(us) | Mlu_io_efficiency | Mlu_compute_efficiency | Mlu_workwpace_size(Bytes) | Data_type | Shape                                                        |
| -------------------------- | --------------------- | ---------------------- | ----------------- | ---------------------- | ------------------------- | --------- | ------------------------------------------------------------ |
| moe_dispatch_backward_gate | 77                    | 16.235                 | 26.9649%          | 5.327%                 | -1                        | float     | samples:4608, hidden:1024, capacity:2880, num_experts: 2     |
| moe_dispatch_backward_gate | 300                   | 7.549                  | 13.86%            | 2.73867%               | -1                        | float     | samples: 18432, hidden: 512, capacity: 11520, num_experts: 2 |

Platform：MLU370

| Operation                  | Mlu_hardware_time(us) | Mlu_interface_time(us) | Mlu_io_efficiency | Mlu_compute_efficiency | Mlu_workwpace_size(Bytes) | Data_type | Shape                                                        |
| -------------------------- | --------------------- | ---------------------- | ----------------- | ---------------------- | ------------------------- | --------- | ------------------------------------------------------------ |
| moe_dispatch_backward_gate | 130                   | 55.04                  | 1.06477           | 0.0709962              | -1                        | float     | samples: 4608, hidden: 1024, capacity: 2880, num_experts: 2  |
| moe_dispatch_backward_gate | 586                   | 14.67                  | 0.473038          | 0.0315461              | -1                        | float     | samples: 18432, hidden: 512, capacity: 11520, num_experts: 2 |

### 3.4 Summary Analysis

- float类型数据测试通过。
- half类型数据，因参考接口实现有bug，暂未测试。

### CODE REVIEW

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+
+- 输入`input` 维度`[samples, hidden]`，输出`grad_gates`的维度`[samples, num_experts]`，因需要在hidden维度进行规约求和，因此拆分`samples`，将`samples`平均拆分到所有task上处理，分两种情况：
+
+  - 当 samples <= taskDim 时，则表示一个sample由多个task处理，拆分伪代码如下：
+
+    ```c
+    // 每个sample由多少个task来计算
+    int one_sample_task_num = taskDim / samples;
+    // 剩余task数，可将剩余task均分给前n个sample
+    int rem_task = taskDim / samples;
```
Comment (starkhu) [inline docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:499]: rem_task应该是taskDim % samples ？

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+       } else {
+           sample_idx = (int)((taskId - rem_task) / one_sample_task_num);
+       }
+       
+       // 根据tadkId，计算需要处理的hidden的大小及偏移
+       int logic_tid = taskId % one_sample_task_num;
+       int hidden_per_task = hidden / one_sample_task_num;
+       int rem_hidden_num = hidden % one_sample_task_num;
+       int hidden_seg_num = hidden_per_task + (int)(logic_tid < rem_hidden_num);
+       int hidden_data_offet = logic_tid * hidden_per_task + ((logic_tid < rem_hidden_num) ? logic_tid : rem_hidden_num);
```
Comment (starkhu) [inline docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:322]: offet -> offset

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md (hunk 1 above)
Comment (wushaoqiang2168) [reply docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:499]: 写错了

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+  for (int i = 0; i < samples_; ++i) {
+    grad_gates[i] = 0.0;
+    if (locations[i] < 0 || locations[i] >= capacity_ || indices[i] < 0 ||
+        indices[i] >= num_experts_) {
+      continue;
+    }
+    for (int j = 0; j < hidden_; ++j) {
+      grad_gates[i] +=
+          dispatch[(indices[i] * capacity_ + locations[i]) * hidden_ + j] *
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp:117]: 好建议

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md (hunk 2 above)
Comment (wushaoqiang2168) [reply docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:322]: 已修改

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md (hunk 1 above)
Comment (wushaoqiang2168) [reply docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:499]: 已修改

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+}
+#endif
+
+template <typename T>
+__mlu_global__ void MLUKernelMoeDispatchBwdGate1(
+    const int *indices, const int *locations, const T *input, const T *dispatch,
+    const int samples, const int capacity, const int hidden,
+    const int num_experts, T *workspace, T *grad_gates) {
+#if __BANG_ARCH__ >= 322
+  if (coreId == 0x80) {
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:62]: 已修改

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+  // check zero element
+  if (zero_element == true) {
+    VLOG(5) << "[mluOpMoeDispatchBackwardGate] Skip zero element tensor.";
+    return MLUOP_STATUS_SUCCESS;
+  }
+
+  if (MLUOP_GEN_CASE_ON_NEW) {
+    GEN_CASE_START("moe_dispatch_backward_gate");
+    GEN_CASE_HANDLE(handle);
+    GEN_CASE_DATA(true, "indices", indices, indices_desc, 10, 0);
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu:196]: 需要，已修改为保存真值

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "moe_dispatch_backward_gate.h"
+#include <string>
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu:24]:
```suggestion
#include "moe_dispatch_backward_gate.h"

#include <string>
```

File: bangc-ops/kernels/utils/common.h [hunk 7]
```diff
@@ -124,7 +124,7 @@ static __mlu_func__ void computeDiv(T *nram_dst, T *nram_src0, T *nram_src1,
   }
 }
 
-/******************************************************************************************
+/*******************************************************************************
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/utils/common.h:127]: 修改这里的原因是？

File: bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto [hunk 8]
```diff
@@ -241,6 +241,7 @@ message Node {
   optional IndiceConvolutionForwardParam indice_convolution_forward_param = 23020721;  //param
   optional CarafeForwardParam  carafe_forward_param     = 408;  // CarafeForwardParam
   optional CarafeBackwardParam  carafe_backward_param   = 409;  // CarafeBackwardParam
+  optional MoeDispatchBackwardGateParam  moe_dispatch_backward_gate_param   = 135246;  //param
```
Comment (WangYuan-cn) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto:244]:
```suggestion
  optional MoeDispatchBackwardGateParam  moe_dispatch_backward_gate_param   = 135246;  // param
```

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+
+- 2023.2.13-2.17 调研源码+开始设计方案
+- 2023.2.20-2.22 设计方案：算子功能+接口设计
+- 2023.2.23 gtest 代码开发
+- 2023.2.24 generator开发
+- 2023.2.27-3.3 完成host/device 开发
+- 2023.3.6-3.8 批量测试+测试报告
+- 2023.3.9-3.10 提交 PR、算子入库
+
+### 5.2 风险分析
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:781]: 添加 user_guid/9/index.rest 下内容

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "kernels/kernel.h"
+#include "kernels/utils/common.h"
+
+__nram__ char nram_buffer[MAX_NRAM_SIZE];
+
+#if __BANG_ARCH__ >= 322
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:28]: 和host侧防呆对齐，此处应该是 372？

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+    int hidden_data_offset =
+        logic_tid * hidden_per_task +
+        ((logic_tid < rem_hidden_num) ? logic_tid : rem_hidden_num);
+
+    if (hidden_seg_num > 0) {
+      // nram space
+      // ping/pong: |nram_input|nram_dispatch|
+      int max_nram_num = MAX_NRAM_SIZE / sizeof(T);
+      int deal_h = 0;
+      deal_h = max_nram_num / 4;
```
Comment (defei-coder) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:159]: 和上面一行，合并成一行

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+    __bang_mul_scalar(nram_idx, nram_indices, capacity, deal_s_num);
+    __bang_add(nram_idx, nram_idx, nram_location, deal_s_num);
+    __bang_mul_scalar(nram_idx, nram_idx, hidden, deal_s_num);
+
+    // 判断 0 <= nram_location < capacity
+    __bang_ge_scalar(nram_mask, nram_location, (int)0, deal_s_num);
+    __bang_lt_scalar(nram_location, nram_location, capacity, deal_s_num);
+    __bang_and(nram_mask, nram_mask, nram_location, deal_s_num);
+
+    // 判断 0 <= nram_indices < num_experts
```
Comment (defei-coder) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:284]: 删除中文注释

File: bangc-ops/kernels/utils/common.h [hunk 13]
```diff
@@ -269,6 +269,34 @@ static __mlu_func__ void computeSigmoid(T *nram_dst, T *nram_src,
...
+      __bang_sumpool((T *)dst, (T *)dst, low_dim, 1, remain_s, 1, remain_s, 1,
+                     1);
+    }
+    if (repeat_s) {
+      __bang_sumpool((T *)dst + (remain_s > 0 ? low_dim : 0),
+                     (T *)dst + remain_s * low_dim, low_dim,
+                     kernel_limit * repeat_s, 1, kernel_limit, 1, 1,
+                     kernel_limit);
+    }
+    high_dim = repeat_s + (bool)remain_s;
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/kernels/utils/common.h:295]: nl中也是这么写的，不过确实不对，应该要修改下

File: bangc-ops/kernels/utils/common.h (hunk 7 above)
Comment (wushaoqiang2168) [reply bangc-ops/kernels/utils/common.h:127]: 超出了80个字符，换行了

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu (hunk 6 above)
Comment (wushaoqiang2168) [reply bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate.mlu]: 已修改

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+  //       printf("%f,",input[i * hidden_ + j]);
+  //     }
+  // }
+  // printf("\n");
+  // printf("dispatch:\n");
+  // for (int i = 0; i < num_experts_ * capacity_; ++i) {
+  //   for (int j = 0; j < hidden_; ++j) {
+  //     printf("%f,",dispatch[i*hidden_ + j]);
+  //   }
+  // }
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/moe_dispatch_backward_gate/moe_dispatch_backward_gate.cpp:135]: 已修改

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -8788,6 +8788,28 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
                               const mluOpTensorDescriptor_t features_out_desc,
                               void *features_out);
 
+mluOpStatus_t MLUOP_WIN_API
```
[the remark this answers is not in the document]
Comment (wushaoqiang2168) [reply bangc-ops/mlu_op.h:8791]: 随后补充

File: docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md (hunk 9 above)
Comment (wushaoqiang2168) [reply docs/bangc-docs/design_docs/moe_dispatch_backward_gate/moe_dispatch_backward_gate.md:782]: 随后补充

File: bangc-ops/mlu_op_kernel.h [hunk 16]
```diff
@@ -563,6 +563,31 @@ void MLUOP_WIN_API mluOpUnion1KernelThreeNNForwardHalf(
     const int n, const int m, const void *unknown, const void *known,
     void *dist2, void *idx);
 
+void MLUOP_WIN_API mluOpUnionKernelMoeDispatchBwdGate1Half(
```
Comment (tudejiang79) [inline bangc-ops/mlu_op_kernel.h:566]: 少一个注释 /* 算子名字 */

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+// template void MLUKernelMoeDispatchBwdGate2<half>(
+//     const int *indices, const int *locations, const half *input,
+//     const half *dispatch, const int samples, const int capacity,
+//     const int hidden, const int num_experts, half *grad_gates);
+
+// template void MLUKernelMoeDispatchBwdGate2<float>(
+//     const int *indices, const int *locations, const float *input,
+//     const float *dispatch, const int samples, const int capacity,
+//     const int hidden, const int num_experts, float *grad_gates);
+
```
Comment (tudejiang79) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:379]: 以上四个可以删除

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 18]
```diff
@@ -0,0 +1,422 @@
...
+    sample_idx = (int)((taskId - rem_task) / one_sample_task_num);
+  }
+
+  int indice = indices[sample_idx];
+  int location = locations[sample_idx];
+  T gard_gates_temp = (T)0.0;
+
+  if (location >= 0 && location < capacity && indice >= 0 &&
+      indice < num_experts) {
+    int logic_tid = taskId % one_sample_task_num;
```
Comment (tudejiang79) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:146]:
```suggestion
    if (taskId < (one_sample_task_num + 1) * rem_task){
        int logic_tid = taskId % one_sample_task_num;
    }else{
       int logic_tid = (taskId - rem_task) % one_sample_task_num;
    }
    
```
个人感觉这里是不是应该这样，比如 32 个 ipu 5 个 sample , 7 ,7, 6,6,6 当taskId = 15时候，logic_tid = (14-2) % 6 = 0 ,占据当前hidden的第一位置，而不是 logic_tid = 14 % 6 = 2

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 19]
```diff
@@ -0,0 +1,422 @@
...
+    int rem_hidden_num = hidden % one_sample_task_num;
+    int hidden_seg_num = hidden_per_task + (int)(logic_tid < rem_hidden_num);
+    int hidden_data_offset =
+        logic_tid * hidden_per_task +
+        ((logic_tid < rem_hidden_num) ? logic_tid : rem_hidden_num);
+
+    if (hidden_seg_num > 0) {
+      // nram space
+      // ping/pong: |nram_input|nram_dispatch|
+      int max_nram_num = MAX_NRAM_SIZE / sizeof(T);
```
Comment (tudejiang79) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:157]: 这里不要对齐128字节吗

File: bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu [hunk 20]
```diff
@@ -0,0 +1,422 @@
...
+
+    // 判断 0 <= nram_location < capacity
+    __bang_ge_scalar(nram_mask, nram_location, (int)0, deal_s_num);
+    __bang_lt_scalar(nram_location, nram_location, capacity, deal_s_num);
+    __bang_and(nram_mask, nram_mask, nram_location, deal_s_num);
+
+    // 判断 0 <= nram_indices < num_experts
+    __bang_ge_scalar(nram_location, nram_indices, (int)0, deal_s_num);
+    __bang_lt_scalar(nram_indices, nram_indices, num_experts, deal_s_num);
+    __bang_and(nram_mask, nram_mask, nram_location, deal_s_num);
```
Comment (tudejiang79) [inline bangc-ops/kernels/moe_dispatch_backward_gate/moe_dispatch_backward_gate_union1.mlu:287]:
```suggestion
   
```
这行是不是多余

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -8788,6 +8788,164 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
...
+ * tensor \b grad_gates.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * moe_dispatch_backward_gate operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] indices_desc
+ * The descriptor of input tensor \b indices, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] indices
+ * Pointer to the MLU memory that stores the indices tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:8852]: \b indices tensor

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -8788,6 +8788,164 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
...
+ * @param[in] locations_desc
+ * The descriptor of input tensor \b locations. which contains dimension, data type and data layout.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:8854]: 句号改为逗号, 英文的逗号

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -10002,6 +10002,169 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the \b input tensor.
+ * @param[in] dispatch_desc
+ * The descriptor of input tensor \b dispatch, which contains dimension, data type and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] dispatch
+ * Pointer to the MLU memory that stores the \b dispatch tensor.
+ * @param[in] samples
+ * The input parameter \b samples indicates the number of elements to input tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10082]: The number of elements to input tensor.

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -10002,6 +10002,169 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
...
+ * @param[in] capacity
+ * The input parameter \b capacity indicates the maximum number of inputs that experts can process.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10084]: The maximum number of inputs that experts can process.

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -10002,6 +10002,169 @@ mluOpIndiceConvolutionForward(mluOpHandle_t handle,
...
+ * @param[in] hidden
+ * The input parameter \b hidden indicates the vector length of a single token.
+ * @param[in] num_experts
+ * The input parameter \b num_experts indicates the number of experts.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10088]: The number of experts.

**Review by GraceGuanH — APPROVED** [review]

**Review by tudejiang79 — APPROVED** [review]

**Review by defei-coder — APPROVED** [review]

**Pull request merged**